031-custom_authentication.md

```diff
@@ -0,0 +1,96 @@
+
+However, there is no ability to specify custom authentication, which is what this proposal will focus on.
+
+## Motivation
+
+Supporting custom authentication allows greater flexibility for users of Strimzi who may have bespoke/proprietary authentication requirements. This would greatly broaden user adoption of the operator, as it would also support more common authentication schemes, such as [Amazon MSK](https://docs.aws.amazon.com/msk/latest/developerguide/security_iam_service-with-iam.html), and attract users who have much more stringent requirements (i.e. proprietary auth). 
+
+The “cost” of supporting such feature should be minimal as well, as the expectation for custom authentication passes the complexity onto the user. Strimzi should strictly support the passing of configuration values, and nothing else. It’s up to the user to ensure their authn/z schemes can be integrated with Strimzi. Considering this also requires custom built Strimzi Kafka images in order to load custom classes on to the classpath, it’s also expected that will only really apply to highly technical/pro users.
+
+## Proposal
+
+Add a CustomAuthenticationKafkaListener class, which supports an example of the following properties
+
+```
+listener.name.<listener-name>.oauthbearer.sasl.client.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.sasl.server.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.sasl.login.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.connections.max.reauth.ms=
+listener.name.<listener-name>.sasl.enabled.mechanisms=
+listener.name.<listener-name>.oauthbearer.sasl.jaas.config=
+listener.name.<listener-name>.AWS_MSK_IAM.sasl.jaas.config=
+listener.name.<listener-name>.AWS_MSK_IAM.sasl.client.callback.handler.class=
+security.protocol=
+principal.builder.class=
+```
+
+Really, this can be boiled down to: 
+
+```
+listener.name.<lister-name>.<auth-type>.*
+listener.name.<listener-name>.sasl.enabled.mechanisms
+security.protocol
+principal.builder.class
+```
+
+Taking this further, the desired yaml should look like: 
+
+```
+    listeners:
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+        authentication:
+          type: custom
+          principal.builder.class: SimplePrincipal.class
+          security.protocol: SASL_SSL
+          sasl.enabled.mechanisms:
+            - OAUTHBEARER
+          listener-config:
+            oauthbearer.sasl.client.callback.handler.class: client.class
+            oauthbearer.sasl.server.callback.handler.class: server.class
+            oauthbearer.sasl.login.callback.handler.class: login.class
+            oauthbearer.connections.max.reauth.ms: 999999999
```

Comment:
I think this should all be in some `config` section which I guess should be just a `Map<String, String>` because each connector will need different options.

Reply:
Sure, that works. One thing to note: `security.protocol: SASL_SSL` is going to be a bit tricky, as we need to append this to `listener.security.protocol.map`, which is why I was thinking of having it be a separate property. 

Also, should have made this a bit more clear, but listener-config I was intending on being a `Map<String, String>`, however  we would forcefully prefix all config settings with `listener.name.<listener-name>` to prevent users from manually patching in forbidden config prefixes.